File: mobile/src/app/pages/trainer/templates/trainer-templates.page.ts
import { Component, OnInit, inject } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { AlertController, ToastController } from '@ionic/angular';
import {
  IonBackButton,
  IonButton,
  IonButtons,
  IonContent,
  IonHeader,
  IonIcon,
  IonRefresher,
  IonRefresherContent,
  IonTitle,
  IonToolbar
} from '@ionic/angular/standalone';
import { addIcons } from 'ionicons';
import { addOutline, trashOutline, createOutline } from 'ionicons/icons';

import {
  StandardTemplateRecord,
  TemplateField,
  TemplatePayload,
  TemplatesApiService,
  TrackingTemplateRecord
} from '../../../core/api/templates-api.service';

/** Tracking template library: create/edit with a field builder, adopt standards, delete. */
@Component({
  selector: 'app-trainer-templates',
  standalone: true,
  imports: [
    FormsModule,
    IonHeader,
    IonToolbar,
    IonTitle,
    IonButtons,
    IonBackButton,
    IonButton,
    IonIcon,
    IonContent,
    IonRefresher,
    IonRefresherContent
  ],
  template: `
    <ion-header>
      <ion-toolbar>
        <ion-buttons slot="start"><ion-back-button defaultHref="/trainer/tabs/manage" /></ion-buttons>
        <ion-title>Templates</ion-title>
        <ion-buttons slot="end">
          <ion-button (click)="startCreate()" aria-label="New template"><ion-icon slot="icon-only" name="add-outline" /></ion-button>
        </ion-buttons>
      </ion-toolbar>
    </ion-header>
    <ion-content>
      <ion-refresher slot="fixed" (ionRefresh)="refresh($event)">
        <ion-refresher-content />
      </ion-refresher>
      <div class="page-pad">
        <div class="kpi-grid">
          <div class="kpi-tile"><span>Templates used</span><strong>{{ templates.length }} / {{ maxTemplates || '—' }}</strong></div>
          <div class="kpi-tile"><span>Assigned clients</span><strong>{{ assignedTotal }}</strong></div>
        </div>

        @if (message) {
          <p class="error-text">{{ message }}</p>
        }

        @if (editorOpen) {
          <div class="card">
            <h3>{{ editingId ? 'Edit template' : 'New template' }}</h3>
            <div class="form-grid">
              <label><span>Name</span><input [(ngModel)]="draft.name" placeholder="e.g. Daily Nutrition Log" /></label>
              <label><span>Purpose</span><input [(ngModel)]="draft.purpose" placeholder="What does this track?" /></label>
              <div class="form-two">
                <label>
                  <span>Cadence</span>
                  <select [(ngModel)]="draft.cadence">
                    <option value="daily">Daily</option>
                    <option value="weekly">Weekly</option>
                    <option value="monthly">Monthly</option>
                  </select>
                </label>
                <label><span>Accent color</span><input type="color" [(ngModel)]="draft.accent" style="height:2.6rem;padding:.2rem" /></label>
              </div>
            </div>

            <p class="hint-note" style="margin-top:.6rem">Fields</p>
            @for (field of draft.custom_fields; track $index) {
              <div style="border:1px dashed var(--app-border);border-radius:.7rem;padding:.7rem;margin-top:.5rem">
                <div class="form-two">
                  <label><span style="display:block;margin-bottom:.25rem;color:var(--app-muted);font-size:.72rem;font-weight:800;text-transform:uppercase">Label</span><input [(ngModel)]="field.label" placeholder="e.g. Weight (kg)" style="width:100%;border:1px solid var(--app-border);border-radius:.6rem;padding:.55rem .7rem;background:var(--app-surface);color:var(--app-text)" /></label>
                  <label><span style="display:block;margin-bottom:.25rem;color:var(--app-muted);font-size:.72rem;font-weight:800;text-transform:uppercase">Type</span>
                    <select [(ngModel)]="field.field_type" style="width:100%;border:1px solid var(--app-border);border-radius:.6rem;padding:.55rem .7rem;background:var(--app-surface);color:var(--app-text)">
                      <option value="number">Number</option>
                      <option value="short_text">Short text</option>
                      <option value="long_text">Long text</option>
                      <option value="yes_no">Yes / No</option>
                      <option value="dropdown">Dropdown</option>
                      <option value="rating">Rating</option>
                    </select>
                  </label>
                </div>
                @if (field.field_type === 'dropdown') {
                  <label style="display:block;margin-top:.4rem">
                    <span style="display:block;margin-bottom:.25rem;color:var(--app-muted);font-size:.72rem;font-weight:800;text-transform:uppercase">Options (comma separated)</span>
                    <input [ngModel]="(field.options || []).join(', ')" (ngModelChange)="setOptions(field, $event)" style="width:100%;border:1px solid var(--app-border);border-radius:.6rem;padding:.55rem .7rem;background:var(--app-surface);color:var(--app-text)" />
                  </label>
                }
                @if (field.field_type === 'rating') {
                  <label style="display:block;margin-top:.4rem">
                    <span style="display:block;margin-bottom:.25rem;color:var(--app-muted);font-size:.72rem;font-weight:800;text-transform:uppercase">Scale (2–10)</span>
                    <input type="number" min="2" max="10" [(ngModel)]="field.scale" style="width:100%;border:1px solid var(--app-border);border-radius:.6rem;padding:.55rem .7rem;background:var(--app-surface);color:var(--app-text)" />
                  </label>
                }
                <div style="text-align:right;margin-top:.3rem">
                  <ion-button size="small" fill="clear" color="danger" (click)="draft.custom_fields.splice($index, 1)">
                    <ion-icon slot="icon-only" name="trash-outline" />
                  </ion-button>
                </div>
              </div>
            }
            <ion-button size="small" fill="outline" style="margin-top:.6rem" (click)="addField()">
              <ion-icon slot="start" name="add-outline" />Add field
            </ion-button>
            <div class="form-two" style="margin-top:.8rem">
              <ion-button size="small" (click)="save()" [disabled]="isSaving || !draft.name.trim() || !draft.custom_fields.length">
                {{ isSaving ? 'Saving…' : 'Save template' }}
              </ion-button>
              <ion-button size="small" fill="clear" (click)="editorOpen = false">Cancel</ion-button>
            </div>
          </div>
        }

        <div class="section-row"><h2>Tracking Library</h2></div>
        <div class="row-list">
          @for (template of templates; track template.id) {
            <div class="row-item">
              <span style="width:.45rem;height:2.6rem;border-radius:1rem;flex:0 0 auto" [style.background]="template.accent || 'var(--app-primary)'"></span>
              <div class="row-main">
                <h3>{{ template.name }}</h3>
                <p>{{ template.cadence }} · {{ template.fields.length }} fields · {{ template.assigned_count || 0 }} assigned</p>
              </div>
              <div class="row-side" style="display:flex;gap:.15rem">
                <ion-button size="small" fill="clear" (click)="startEdit(template)"><ion-icon slot="icon-only" name="create-outline" /></ion-button>
                <ion-button size="small" fill="clear" color="danger" (click)="remove(template)"><ion-icon slot="icon-only" name="trash-outline" /></ion-button>
              </div>
            </div>
          } @empty {
            <p class="empty-note">No templates yet. Create one or adopt a standard template below.</p>
          }
        </div>

        @if (unAdoptedStandards.length) {
          <div class="section-row"><h2>Standard templates</h2></div>
          <div class="row-list">
            @for (standard of unAdoptedStandards; track standard.key) {
              <div class="row-item">
                <div class="row-main">
                  <h3>{{ standard.name }}</h3>
                  <p>{{ standard.cadence }} · {{ standard.fields.length }} fields</p>
                </div>
                <div class="row-side">
                  <ion-button size="small" fill="outline" (click)="adopt(standard)">Adopt</ion-button>
                </div>
              </div>
            }
          </div>
        }
        <div class="bottom-space"></div>
      </div>
    </ion-content>
  `
})
export class TrainerTemplatesPage implements OnInit {
  private readonly api = inject(TemplatesApiService);
  private readonly alertController = inject(AlertController);
  private readonly toastController = inject(ToastController);

  templates: TrackingTemplateRecord[] = [];
  standards: StandardTemplateRecord[] = [];
  maxTemplates = 0;
  message = '';
  editorOpen = false;
  editingId = 0;
  isSaving = false;
  draft: TemplatePayload = { name: '', purpose: '', cadence: 'daily', accent: '#0b7de3', custom_fields: [] };

  constructor() {
    addIcons({ addOutline, trashOutline, createOutline });
  }

  get assignedTotal(): number {
    return this.templates.reduce((sum, item) => sum + Number(item.assigned_count || 0), 0);
  }

  get unAdoptedStandards(): StandardTemplateRecord[] {
    return this.standards.filter((standard) => !standard.adopted);
  }

  ngOnInit(): void {
    this.load();
  }

  refresh(event: CustomEvent): void {
    this.load(() => (event.target as HTMLIonRefresherElement).complete());
  }

  startCreate(): void {
    this.editingId = 0;
    this.draft = { name: '', purpose: '', cadence: 'daily', accent: '#0b7de3', custom_fields: [] };
    this.addField();
    this.editorOpen = true;
  }

  startEdit(template: TrackingTemplateRecord): void {
    this.editingId = template.id;
    this.draft = {
      name: template.name,
      purpose: template.purpose,
      cadence: template.cadence,
      accent: template.accent || '#0b7de3',
      custom_fields: template.fields.map((field) => ({ ...field }))
    };
    this.editorOpen = true;
  }

  addField(): void {
    this.draft.custom_fields.push({ label: '', field_type: 'number', placeholder: '' });
  }

  setOptions(field: TemplateField, raw: string): void {
    field.options = raw
      .split(',')
      .map((option) => option.trim())
      .filter(Boolean);
  }

  save(): void {
    this.isSaving = true;
    const payload: TemplatePayload = {
      ...this.draft,
      name: this.draft.name.trim(),
      purpose: this.draft.purpose.trim(),
      custom_fields: this.draft.custom_fields.filter((field) => field.label.trim())
    };
    const request = this.editingId ? this.api.updateTemplate(this.editingId, payload) : this.api.createTemplate(payload);

    request.subscribe({
      next: () => {
        this.isSaving = false;
        this.editorOpen = false;
        this.load();
        void this.toast('Template saved.');
      },
      error: () => {
        this.isSaving = false;
        this.message = 'Could not save the template.';
      }
    });
  }

  adopt(standard: StandardTemplateRecord): void {
    this.api.adoptStandardTemplate(standard.key).subscribe({
      next: () => {
        this.load();
        void this.toast(`${standard.name} added.`);
      },
      error: () => (this.message = 'Could not adopt the template.')
    });
  }

  async remove(template: TrackingTemplateRecord): Promise<void> {
    const alert = await this.alertController.create({
      header: `Delete ${template.name}?`,
      message: 'Assigned clients lose access to this template. Past entries stay saved.',
      buttons: [
        { text: 'Cancel', role: 'cancel' },
        {
          text: 'Delete',
          role: 'destructive',
          handler: () => {
            this.api.deleteTemplate(template.id).subscribe({
              next: () => this.load(),
              error: () => (this.message = 'Template could not be deleted.')
            });
          }
        }
      ]
    });
    await alert.present();
  }

  private load(done?: () => void): void {
    this.api.getTemplates().subscribe({
      next: (response) => {
        this.templates = response.templates;
        this.maxTemplates = response.max_templates;
        this.message = '';
        done?.();
      },
      error: () => {
        this.message = 'Could not load templates.';
        done?.();
      }
    });
    this.api.getStandardTemplates().subscribe({
      next: (response) => (this.standards = response.standard_templates),
      error: () => (this.standards = [])
    });
  }

  private async toast(text: string): Promise<void> {
    const toast = await this.toastController.create({ message: text, duration: 1800, position: 'bottom' });
    await toast.present();
  }
}

File: mobile_flutter/lib/features/trainer/trainer_templates_page.dart
import 'package:flutter/material.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';
import 'package:go_router/go_router.dart';

import '../../app/router.dart';
import '../../core/api/api_client.dart';
import '../../core/api/models/template_models.dart';
import '../../core/api/templates_api.dart';
import '../../core/theme/app_tokens.dart';
import '../../shared/widgets/app_widgets.dart';

/// Tracking template library: create/edit with a field builder, adopt
/// standards, delete.
/// Replica of mobile/src/app/pages/trainer/templates/trainer-templates.page.ts.
class TrainerTemplatesPage extends ConsumerStatefulWidget {
  const TrainerTemplatesPage({super.key});

  @override
  ConsumerState<TrainerTemplatesPage> createState() =>
      _TrainerTemplatesPageState();
}

/// Mutable field row for the builder — TemplateField is immutable.
class _DraftField {
  _DraftField({
    this.label = '',
    this.fieldType = TemplateFieldType.number,
    this.placeholder = '',
    List<String>? options,
    this.scale,
  }) : options = options ?? [];

  String label;
  String fieldType;
  String placeholder;
  List<String> options;
  int? scale;

  TemplateField toField() => TemplateField(
        label: label.trim(),
        fieldType: fieldType,
        placeholder: placeholder,
        options: fieldType == TemplateFieldType.dropdown ? options : const [],
        scale: fieldType == TemplateFieldType.rating ? (scale ?? 5) : null,
      );

  static _DraftField from(TemplateField field) => _DraftField(
        label: field.label,
        fieldType: field.fieldType,
        placeholder: field.placeholder,
        options: [...field.options],
        scale: field.scale,
      );
}

class _TrainerTemplatesPageState extends ConsumerState<TrainerTemplatesPage> {
  List<TrackingTemplateRecord> _templates = [];
  List<StandardTemplateRecord> _standards = [];
  int _maxTemplates = 0;
  String _message = '';
  bool _loading = true;

  bool _editorOpen = false;
  int _editingId = 0;
  bool _isSaving = false;

  final _name = TextEditingController();
  final _purpose = TextEditingController();
  String _cadence = TemplateCadence.daily;
  Color _accent = const Color(0xFF0B7DE3);
  List<_DraftField> _fields = [];

  @override
  void initState() {
    super.initState();
    _load();
  }

  @override
  void dispose() {
    _name.dispose();
    _purpose.dispose();
    super.dispose();
  }

  int get _assignedTotal =>
      _templates.fold(0, (sum, item) => sum + item.assignedCount);

  List<StandardTemplateRecord> get _unAdoptedStandards =>
      _standards.where((standard) => !standard.adopted).toList();

  bool get _canSave =>
      !_isSaving &&
      _name.text.trim().isNotEmpty &&
      _fields.any((field) => field.label.trim().isNotEmpty);

  Future<void> _load() async {
    try {
      final response = await ref.read(templatesApiProvider).getTemplates();
      if (!mounted) return;
      setState(() {
        _templates = response.templates;
        _maxTemplates = response.maxTemplates;
        _message = '';
        _loading = false;
      });
    } catch (_) {
      if (!mounted) return;
      setState(() {
        _message = 'Could not load templates.';
        _loading = false;
      });
    }
    try {
      final standards = await ref.read(templatesApiProvider).getStandardTemplates();
      if (mounted) setState(() => _standards = standards);
    } catch (_) {
      if (mounted) setState(() => _standards = []);
    }
  }

  void _startCreate() {
    setState(() {
      _editingId = 0;
      _name.clear();
      _purpose.clear();
      _cadence = TemplateCadence.daily;
      _accent = const Color(0xFF0B7DE3);
      _fields = [_DraftField()];
      _editorOpen = true;
    });
  }

  void _startEdit(TrackingTemplateRecord template) {
    setState(() {
      _editingId = template.id;
      _name.text = template.name;
      _purpose.text = template.purpose;
      _cadence = template.cadence.isEmpty ? TemplateCadence.daily : template.cadence;
      _accent = _parseAccent(template.accent);
      _fields = template.fields.map(_DraftField.from).toList();
      _editorOpen = true;
    });
  }

  /// Backend stores accents as "#rrggbb"; fall back to brand blue.
  Color _parseAccent(String hex) {
    final cleaned = hex.replaceAll('#', '').trim();
    if (cleaned.length != 6) return const Color(0xFF0B7DE3);
    final value = int.tryParse(cleaned, radix: 16);
    return value == null ? const Color(0xFF0B7DE3) : Color(0xFF000000 | value);
  }

  String _accentHex(Color color) =>
      '#${(color.toARGB32() & 0xFFFFFF).toRadixString(16).padLeft(6, '0')}';

  Future<void> _save() async {
    setState(() => _isSaving = true);
    final payload = TemplatePayload(
      name: _name.text.trim(),
      purpose: _purpose.text.trim(),
      cadence: _cadence,
      accent: _accentHex(_accent),
      customFields: _fields
          .where((field) => field.label.trim().isNotEmpty)
          .map((field) => field.toField())
          .toList(),
    );

    final api = ref.read(templatesApiProvider);
    try {
      if (_editingId > 0) {
        await api.updateTemplate(_editingId, payload);
      } else {
        await api.createTemplate(payload);
      }
      if (!mounted) return;
      setState(() {
        _isSaving = false;
        _editorOpen = false;
      });
      await _load();
      if (mounted) _toast('Template saved.');
    } on ApiException catch (error) {
      if (!mounted) return;
      setState(() {
        _isSaving = false;
        _message = error.message;
      });
    }
  }

  Future<void> _adopt(StandardTemplateRecord standard) async {
    try {
      await ref.read(templatesApiProvider).adoptStandardTemplate(standard.key);
      await _load();
      if (mounted) _toast('${standard.name} added.');
    } on ApiException catch (error) {
      if (mounted) setState(() => _message = error.message);
    }
  }

  Future<void> _remove(TrackingTemplateRecord template) async {
    // A template still assigned to clients cannot be deleted: the trainer must
    // unassign it from each client first. This protects clients from silently
    // losing a tracker they are actively logging against.
    if (template.assignedCount > 0) {
      await showDialog<void>(
        context: context,
        builder: (context) => AlertDialog(
          title: Text('${template.name} is in use'),
          content: Text(
            'This template is assigned to ${template.assignedCount} '
            '${template.assignedCount == 1 ? 'client' : 'clients'}.\n\n'
            'Remove it from every client first, then you can delete it.',
          ),
          actions: [
            TextButton(
              onPressed: () => context.pop(),
              child: const Text('Close'),
            ),
            FilledButton(
              onPressed: () {
                context.pop();
                context.go(Routes.trainerClients);
              },
              style: FilledButton.styleFrom(
                minimumSize: const Size(0, AppSize.buttonHeightSm),
              ),
              child: const Text('Open Clients'),
            ),
          ],
        ),
      );
      return;
    }

    final confirmed = await showDialog<bool>(
      context: context,
      builder: (context) => AlertDialog(
        title: Text('Delete ${template.name}?'),
        content: const Text(
          'This template is not assigned to anyone. Past entries stay saved.',
        ),
        actions: [
          TextButton(
            onPressed: () => context.pop(false),
            child: const Text('Cancel'),
          ),
          FilledButton(
            onPressed: () => context.pop(true),
            style: FilledButton.styleFrom(
              backgroundColor: context.colors.error,
              minimumSize: const Size(0, AppSize.buttonHeightSm),
            ),
            child: const Text('Delete'),
          ),
        ],
      ),
    );
    if (confirmed != true) return;

    try {
      await ref.read(templatesApiProvider).deleteTemplate(template.id);
      await _load();
    } on ApiException catch (error) {
      if (mounted) setState(() => _message = error.message);
    }
  }

  void _toast(String text) {
    ScaffoldMessenger.of(context).showSnackBar(
      SnackBar(content: Text(text), duration: const Duration(milliseconds: 1800)),
    );
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: const Text('Templates'),
        leading: BackButton(onPressed: () => context.go(Routes.trainerManage)),
        actions: [
          IconButton(
            onPressed: _startCreate,
            icon: const Icon(Icons.add),
            tooltip: 'New template',
          ),
        ],
      ),
      body: PagePad(
        onRefresh: _load,
        children: [
          KpiGrid(
            children: [
              KpiTile(
                label: 'Templates used',
                value: '${_templates.length} / ${_maxTemplates > 0 ? _maxTemplates : '—'}',
              ),
              KpiTile(label: 'Assigned clients', value: '$_assignedTotal'),
            ],
          ),
          if (_message.isNotEmpty) ...[
            const SizedBox(height: AppSpacing.md),
            ErrorNote(message: _message, onRetry: _load),
          ],

          if (_editorOpen) ...[
            const SizedBox(height: AppSpacing.md),
            _Editor(
              editing: _editingId > 0,
              name: _name,
              purpose: _purpose,
              cadence: _cadence,
              accent: _accent,
              fields: _fields,
              isSaving: _isSaving,
              canSave: _canSave,
              onCadence: (value) => setState(() => _cadence = value),
              onAccent: (value) => setState(() => _accent = value),
              onChanged: () => setState(() {}),
              onAddField: () => setState(() => _fields.add(_DraftField())),
              onRemoveField: (index) => setState(() => _fields.removeAt(index)),
              onSave: _save,
              onCancel: () => setState(() => _editorOpen = false),
            ),
          ],

          const SectionHeader(title: 'Tracking Library'),
          if (_loading)
            for (var i = 0; i < 3; i++) const SkeletonBox(height: 66)
          else if (_templates.isEmpty)
            const EmptyState(
              message: 'No templates yet. Create one or adopt a standard below.',
            )
          else
            for (final template in _templates)
              RowItem(
                title: template.name,
                subtitle:
                    '${TemplateCadence.label(template.cadence)} · ${template.fields.length} fields · ${template.assignedCount} assigned',
                leading: Container(
                  width: 5,
                  height: 40,
                  decoration: BoxDecoration(
                    color: _parseAccent(template.accent),
                    borderRadius: BorderRadius.circular(999),
                  ),
                ),
                trailing: Row(
                  mainAxisSize: MainAxisSize.min,
                  children: [
                    IconButton(
                      onPressed: () => _startEdit(template),
                      icon: const Icon(Icons.edit_outlined),
                      iconSize: AppSize.iconRow,
                      tooltip: 'Edit',
                      visualDensity: VisualDensity.compact,
                    ),
                    IconButton(
                      onPressed: () => _remove(template),
                      icon: Icon(
                        template.assignedCount > 0
                            ? Icons.lock_outline
                            : Icons.delete_outline,
                      ),
                      iconSize: AppSize.iconRow,
                      // Locked reads as unavailable, not as a destructive action.
                      color: template.assignedCount > 0
                          ? context.tokens.muted
                          : context.colors.error,
                      tooltip: template.assignedCount > 0
                          ? 'Assigned to ${template.assignedCount} — unassign first'
                          : 'Delete',
                      visualDensity: VisualDensity.compact,
                    ),
                  ],
                ),
              ),

          if (_unAdoptedStandards.isNotEmpty) ...[
            const SectionHeader(title: 'Standard templates'),
            for (final standard in _unAdoptedStandards)
              RowItem(
                title: standard.name,
                subtitle:
                    '${TemplateCadence.label(standard.cadence)} · ${standard.fields.length} fields',
                trailing: OutlinedButton(
                  onPressed: () => _adopt(standard),
                  style: OutlinedButton.styleFrom(
                    minimumSize: const Size(72, AppSize.buttonHeightSm),
                  ),
                  child: const Text('Adopt'),
                ),
              ),
          ],
        ],
      ),
    );
  }
}

class _Editor extends StatelessWidget {
  const _Editor({
    required this.editing,
    required this.name,
    required this.purpose,
    required this.cadence,
    required this.accent,
    required this.fields,
    required this.isSaving,
    required this.canSave,
    required this.onCadence,
    required this.onAccent,
    required this.onChanged,
    required this.onAddField,
    required this.onRemoveField,
    required this.onSave,
    required this.onCancel,
  });

  final bool editing;
  final TextEditingController name;
  final TextEditingController purpose;
  final String cadence;
  final Color accent;
  final List<_DraftField> fields;
  final bool isSaving;
  final bool canSave;
  final ValueChanged<String> onCadence;
  final ValueChanged<Color> onAccent;
  final VoidCallback onChanged;
  final VoidCallback onAddField;
  final ValueChanged<int> onRemoveField;
  final VoidCallback onSave;
  final VoidCallback onCancel;

  /// A native colour picker is a heavy dependency for one field; the Ionic page
  /// used <input type="color">. A swatch row covers the real use (a visual tag)
  /// with far less weight.
  static const _swatches = [
    Color(0xFF0B7DE3), Color(0xFF2563EB), Color(0xFF0F9F9D), Color(0xFF159567),
    Color(0xFFD97706), Color(0xFFD92D20), Color(0xFF7C3AED), Color(0xFF64748B),
  ];

  @override
  Widget build(BuildContext context) {
    final tokens = context.tokens;

    return AppCard(
      child: Column(
        crossAxisAlignment: CrossAxisAlignment.start,
        children: [
          Text(
            editing ? 'Edit template' : 'New template',
            style: context.text.titleSmall,
          ),
          const SizedBox(height: AppSpacing.md),
          TextField(
            controller: name,
            onChanged: (_) => onChanged(),
            decoration: const InputDecoration(
              labelText: 'Name',
              hintText: 'e.g. Daily Nutrition Log',
            ),
          ),
          const SizedBox(height: AppSpacing.md),
          TextField(
            controller: purpose,
            decoration: const InputDecoration(
              labelText: 'Purpose',
              hintText: 'What does this track?',
            ),
          ),
          const SizedBox(height: AppSpacing.md),
          DropdownButtonFormField<String>(
            initialValue: cadence,
            decoration: const InputDecoration(labelText: 'Cadence'),
            items: TemplateCadence.all
                .map((c) => DropdownMenuItem(
                      value: c,
                      child: Text(TemplateCadence.label(c)),
                    ))
                .toList(),
            onChanged: (value) => onCadence(value ?? TemplateCadence.daily),
          ),
          const SizedBox(height: AppSpacing.md),
          Text(
            'ACCENT COLOR',
            style: context.text.labelSmall?.copyWith(
              color: tokens.muted,
              fontWeight: FontWeight.w800,
            ),
          ),
          const SizedBox(height: AppSpacing.sm),
          Wrap(
            spacing: AppSpacing.sm,
            children: [
              for (final swatch in _swatches)
                InkWell(
                  onTap: () => onAccent(swatch),
                  borderRadius: BorderRadius.circular(999),
                  child: Container(
                    width: 30,
                    height: 30,
                    decoration: BoxDecoration(
                      color: swatch,
                      shape: BoxShape.circle,
                      border: Border.all(
                        color: swatch.toARGB32() == accent.toARGB32()
                            ? context.colors.onSurface
                            : Colors.transparent,
                        width: 2.5,
                      ),
                    ),
                  ),
                ),
            ],
          ),

          const SizedBox(height: AppSpacing.lg),
          Text(
            'FIELDS',
            style: context.text.labelSmall?.copyWith(
              color: tokens.muted,
              fontWeight: FontWeight.w800,
            ),
          ),
          for (var i = 0; i < fields.length; i++)
            _FieldEditor(
              field: fields[i],
              onChanged: onChanged,
              onRemove: fields.length > 1 ? () => onRemoveField(i) : null,
            ),
          const SizedBox(height: AppSpacing.sm),
          OutlinedButton.icon(
            onPressed: onAddField,
            icon: const Icon(Icons.add),
            label: const Text('Add field'),
            style: OutlinedButton.styleFrom(
              minimumSize: const Size(0, AppSize.buttonHeightSm),
            ),
          ),
          const SizedBox(height: AppSpacing.md),
          Row(
            children: [
              Expanded(
                child: FilledButton(
                  onPressed: canSave ? onSave : null,
                  child: Text(isSaving ? 'Saving…' : 'Save template'),
                ),
              ),
              const SizedBox(width: AppSpacing.sm),
              TextButton(onPressed: onCancel, child: const Text('Cancel')),
            ],
          ),
        ],
      ),
    );
  }
}

class _FieldEditor extends StatelessWidget {
  const _FieldEditor({
    required this.field,
    required this.onChanged,
    this.onRemove,
  });

  final _DraftField field;
  final VoidCallback onChanged;
  final VoidCallback? onRemove;

  @override
  Widget build(BuildContext context) {
    final tokens = context.tokens;

    return Container(
      margin: const EdgeInsets.only(top: AppSpacing.sm),
      padding: const EdgeInsets.all(AppSpacing.md),
      decoration: BoxDecoration(
        border: Border.all(color: tokens.border, style: BorderStyle.solid),
        borderRadius: AppRadius.smAll,
        color: tokens.surfaceSoft,
      ),
      child: Column(
        children: [
          TextFormField(
            initialValue: field.label,
            onChanged: (value) {
              field.label = value;
              onChanged();
            },
            decoration: const InputDecoration(
              labelText: 'Label',
              hintText: 'e.g. Weight (kg)',
            ),
          ),
          const SizedBox(height: AppSpacing.sm),
          DropdownButtonFormField<String>(
            initialValue: field.fieldType,
            decoration: const InputDecoration(labelText: 'Type'),
            items: TemplateFieldType.all
                .map((t) => DropdownMenuItem(
                      value: t,
                      child: Text(TemplateFieldType.label(t)),
                    ))
                .toList(),
            onChanged: (value) {
              field.fieldType = value ?? TemplateFieldType.number;
              onChanged();
            },
          ),
          if (field.fieldType == TemplateFieldType.dropdown) ...[
            const SizedBox(height: AppSpacing.sm),
            TextFormField(
              initialValue: field.options.join(', '),
              onChanged: (value) {
                field.options = value
                    .split(',')
                    .map((option) => option.trim())
                    .where((option) => option.isNotEmpty)
                    .toList();
              },
              decoration: const InputDecoration(
                labelText: 'Options',
                helperText: 'Comma separated',
              ),
            ),
          ],
          if (field.fieldType == TemplateFieldType.rating) ...[
            const SizedBox(height: AppSpacing.sm),
            TextFormField(
              initialValue: '${field.scale ?? 5}',
              keyboardType: TextInputType.number,
              onChanged: (value) {
                // The engine clamps 2..10 anyway; keep the input honest here too.
                field.scale = int.tryParse(value)?.clamp(2, 10);
              },
              decoration: const InputDecoration(
                labelText: 'Scale',
                helperText: '2 to 10',
              ),
            ),
          ],
          if (onRemove != null)
            Align(
              alignment: Alignment.centerRight,
              child: TextButton.icon(
                onPressed: onRemove,
                icon: const Icon(Icons.delete_outline, size: AppSize.iconRow),
                label: const Text('Remove'),
                style: TextButton.styleFrom(
                  foregroundColor: context.colors.error,
                  minimumSize: const Size(0, 32),
                  tapTargetSize: MaterialTapTargetSize.shrinkWrap,
                ),
              ),
            ),
        ],
      ),
    );
  }
}
